src/devices_models/devices/thermal_generation.jl

```diff
@@ -62,189 +63,128 @@ get_variable_binary(::Union{ColdStartVariable, WarmStartVariable, HotStartVariab
+    devices::IS.FlattenIteratorWrapper{V},
+    model::DeviceModel{V, W},
+    X::Type{<:PM.AbstractPowerModel},
+    feedforward::Union{Nothing, AbstractAffectFeedForward}
+) where {
+    V <: PSY.ThermalGen,
+    W <: ThermalCompactDispatch,
+}
+    add_range_constraints!(container, T, U, devices, model, X, feedforward)
+end
+
+function get_limits(device, ::Type{PowerAboveMinimumVariable}, ::Type{<:ThermalCompactDispatch})
+    (
+        min = 0.0,
+        max = PSY.get_active_power_limits(device).max -
+                PSY.get_active_power_limits(device).min,
     )
 end
 
-"""
-This function adds the active power limits of generators when there are no CommitmentVariables
-"""
-function DeviceRangeConstraintSpec(
-    ::Type{<:ActivePowerVariableLimitsConstraint},
-    ::Type{ActivePowerVariable},
-    ::Type{T},
-    ::Type{<:ThermalDispatchNoMin},
-    ::Type{<:PM.AbstractPowerModel},
-    feedforward::Nothing,
-    use_parameters::Bool,
-) where {T <: PSY.ThermalGen}
-    return DeviceRangeConstraintSpec(;
-        range_constraint_spec = RangeConstraintSpec(;
-            constraint_type = ActivePowerVariableLimitsConstraint(),
-            variable_type = ActivePowerVariable(),
-            limits_func = x -> (min = 0.0, max = PSY.get_active_power_limits(x).max),
-            constraint_func = device_range!,
-            constraint_struct = DeviceRangeConstraintInfo,
-            component_type = T,
-        ),
-        custom_optimization_container_func = custom_active_power_constraints!,
+
+# active power limits of generators when there are no CommitmentVariables
+# TODO: handle custom active power constraints for optimization
+function get_limits(device, ::Type{ActivePowerVariableLimitsConstraint}, ::Type{<:ThermalDispatchNoMin}) #  -> Union{Nothing, NamedTuple{(:startup, :shutdown), Tuple{Float64, Float64}}}
+    (min = 0.0, max = PSY.get_active_power_limits(device).max)
+end
+
+# active power limits of generators. Constraint (17) & (18) from PGLIB
+
+function add_constraints!(
+    container::OptimizationContainer,
+    T::Type{<:PowerVariableLimitsConstraint},
+    U::Type{<:VariableType},
+    devices::IS.FlattenIteratorWrapper{V},
+    model::DeviceModel{V, W},
+    X::Type{<:PM.AbstractPowerModel},
+    feedforward::Union{Nothing, AbstractAffectFeedForward}
+) where {
+    V <: PSY.ThermalGen,
+    W <: AbstractThermalDispatchFormulation,
+}
+    add_range_constraints!(container, T, U, devices, model, X, feedforward)
+end
+
+function get_limits(device, ::Type{ActivePowerVariableLimitsConstraint}, ::Type{<:ThermalMultiStartUnitCommitment}) #  -> Union{Nothing, NamedTuple{(:startup, :shutdown), Tuple{Float64, Float64}}}
+    (
+        min = 0.0,
+        max = PSY.get_active_power_limits(device).max - PSY.get_active_power_limits(device).min,
     )
 end
 
-function custom_active_power_constraints!(
+function get_binary_variable_types(::Type{ActivePowerVariableLimitsConstraint}, ::Type{<:ThermalMultiStartUnitCommitment})
```

Comment:
Do we always need to do this? It seems a bit over the top if isn't used that much and a new formulation of thermal would be required to implement this method